builtin/provisioners/puppet/resource_provisioner.go

```diff
@@ -0,0 +1,337 @@
+				Optional: true,
+			},
+		},
+		ApplyFunc:    applyFn,
+		ValidateFunc: validateFn,
+	}
+}
+
+func applyFn(ctx context.Context) (rerr error) {
+	output := ctx.Value(schema.ProvOutputKey).(terraform.UIOutput)
+	state := ctx.Value(schema.ProvRawStateKey).(*terraform.InstanceState)
+	configData := ctx.Value(schema.ProvConfigDataKey).(*schema.ResourceData)
+
+	p, err := decodeConfig(configData)
+	if err != nil {
+		return err
+	}
+
+	p.instanceState = state
+	p.output = output
+
+	if p.OSType == "" {
+		switch connType := state.Ephemeral.ConnInfo["type"]; connType {
+		case "ssh", "":
+			p.OSType = "nix"
+		case "winrm":
+			p.OSType = "windows"
+		default:
+			return fmt.Errorf("Unsupported connection type: %s", connType)
+		}
+	}
+
+	switch p.OSType {
+	case "nix":
+		p.runPuppetAgent = p.nixRunPuppetAgent
+		p.installPuppetAgent = p.nixInstallPuppetAgent
+		p.uploadFile = p.nixUploadFile
+		p.defaultCertname = p.nixDefaultCertname
+		p.isPuppetEnterprise = p.nixIsPuppetEnterprise
+	case "windows":
+		p.runPuppetAgent = p.windowsRunPuppetAgent
+		p.installPuppetAgent = p.windowsInstallPuppetAgent
+		p.uploadFile = p.windowsUploadFile
+		p.UseSudo = false
+		p.defaultCertname = p.windowsDefaultCertname
+		p.isPuppetEnterprise = p.windowsIsPuppetEnterprise
+	default:
+		return fmt.Errorf("Unsupported OS type: %s", p.OSType)
+	}
+
+	comm, err := communicator.New(state)
+	if err != nil {
+		return err
+	}
+
+	retryCtx, cancel := context.WithTimeout(ctx, comm.Timeout())
+	defer cancel()
+
+	err = communicator.Retry(retryCtx, func() error {
+		return comm.Connect(output)
+	})
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := comm.Disconnect(); err != nil {
+			rerr = err
+		}
+	}()
+
+	p.comm = comm
+
+	_, ok := configData.GetOkExists("open_source")
+	if !ok {
+		isPE, err := p.isPuppetEnterprise()
+		if err != nil {
+			return fmt.Errorf("Unable to determine if %s is running Puppet Enterprise: %v", p.Server, err)
+		}
+		p.OpenSource = !isPE
+	}
+
+	if p.OpenSource {
+		p.installPuppetAgent = p.installPuppetAgentOpenSource
+	}
+
+	csrAttrs := new(csrAttributes)
+	csrAttrs.CustomAttributes = make(map[string]string)
+	csrAttrs.ExtensionRequests = make(map[string]string)
+
+	if p.Autosign {
+		if p.Certname == "" {
+			p.Certname, _ = p.defaultCertname()
+		}
+
+		autosignToken, err := p.generateAutosignToken(p.Certname)
+		if err != nil {
+			return fmt.Errorf("Failed to generate an autosign token: %s", err)
+		}
+		csrAttrs.CustomAttributes["challengePassword"] = autosignToken
+	}
+
+	if p.PPRole != "" {
+		csrAttrs.ExtensionRequests["pp_role"] = p.PPRole
+	}
+
+	if err = p.writeCSRAttributes(csrAttrs); err != nil {
+		return fmt.Errorf("Failed to write csr_attributes.yaml: %s", err)
+	}
+
+	if err = p.installPuppetAgent(); err != nil {
+		return err
+	}
+
+	if err = p.runPuppetAgent(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func validateFn(config *terraform.ResourceConfig) (ws []string, es []error) {
```

Comment:
Is this intentionally blank? I believe if you do not require a `validateFn` it can be omitted (both here and line 96).

Reply:
Cool, I've removed it